sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala

```diff
@@ -32,11 +33,10 @@ object SchemaPruning {
     // in the resulting schema may differ from their ordering in the logical relation's
     // original schema
     val mergedSchema = requestedRootFields
-      .map { case root: RootField => StructType(Array(root.field)) }
+      .map { root: RootField => StructType(Array(root.field)) }
       .reduceLeft(_ merge _)
-    val dataSchemaFieldNames = dataSchema.fieldNames.toSet
     val mergedDataSchema =
-      StructType(mergedSchema.filter(f => dataSchemaFieldNames.contains(f.name)))
+      StructType(dataSchema.map(s => mergedSchema.find(_.name.equals(s.name)).getOrElse(s)))
```

Comment:
what's the actual difference? can you give a simple example?

Reply:
It seems we don't prune anything from the root fields now.

Reply:
if this is the case please update the document of this method.

Reply:
```scala
spark.sql(
  """
    |CREATE TABLE t1 (
    |  _col0 INT,
    |  _col1 STRING,
    |  _col2 STRUCT<c1: STRING, c2: STRING, c3: STRING, c4: BIGINT>)
    |USING ORC
    |""".stripMargin)


spark.sql("SELECT _col0, _col2.c1 FROM t1").show
```
Before this PR, the `pruneDataSchema` returns:
```
`_col0` INT,`_col2` STRUCT<`c1`: STRING>
```

After this PR, the `pruneDataSchema` returns:
```
`_col0` INT,`_col1` STRING,`_col2` STRUCT<`c1`: STRING>
```

It only prune nested schemas.

Reply:
What's wrong with the previous behavior? We can't sacrifice performance for all the cases only because the ORC by ordinal case is problematic.

Reply:
It removed the  ``` `_col1` STRING ```. Nested column pruning still works. Please see these test suites: `ParquetSchemaPruningSuite`, `OrcV1SchemaPruningSuite` and `OrcV2SchemaPruningSuite`.

That why it's only a problem in nested column pruning but not column pruning.

Reply:
I don't know the details enough to understand why nested column pruning still works after the change here. @viirya can you take a look?

Reply:
is it because column pruning will be done by other rules so we don't need to consider it here?

Reply:
> is it because column pruning will be done by other rules so we don't need to consider it here?

Yes.

https://github.com/apache/spark/blob/7a5647a93aaea9d1d78d9262e24fc8c010db04d0/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L208-L213

https://github.com/apache/spark/blob/e64eb75aede71a5403a4d4436e63b1fcfdeca14d/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownUtils.scala#L96-L97

Reply:
Can you provide the full code workflow to explain why this causes issues in ORC? I'm still not very sure.

Reply:
1. Prune nested schema:
https://github.com/apache/spark/blob/0f2c0b53e8fb18c86c67b5dd679c006db93f94a5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala#L28-L43

2. Use this pruned nested schema to build the `dataSchema` in `Relation`
https://github.com/apache/spark/blob/25e7d1ceee8c9f4ecb4ab796f51e9bcbc0500fae/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala#L81-L86

2. The `readDataColumns` is the complete column pruning:
https://github.com/apache/spark/blob/7a5647a93aaea9d1d78d9262e24fc8c010db04d0/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L208-L226

3. `dataSchema` from `relation.dataSchema`. It is the pruned nested schema:
https://github.com/apache/spark/blob/935aa8c8db6824648483f26c2889c33030985259/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala#L398-L407

4. `OrcUtils.requestedColumnIds` use this pruned nested schema:
https://github.com/apache/spark/blob/1fc66f68703e0b14e03fe0ed5ca93e9af20f41a9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala#L193-L197

Reply:
It is because `requestedColumnIds` will check if given data schema has less fields than physical schema in ORC file.

Under nested column pruning, Spark will let data source use pruned schema as data schema to read files. E.g., Spark prune `_col1`, for the above example. But the ORC file has three top-level fields `_col0`, `_col1`, and `_col2`, so the check in `requestedColumnIds` will fail on the case.

Reply:
> > is it because column pruning will be done by other rules so we don't need to consider it here?
> 
> Yes.
> 

Hmm?  In `PushDownUtils.pruneColumns`, if you enable nested column pruning, Spark will only run the path of nested column pruning, not the quoted L96-97.